### PR TITLE
Fix/sequence length power of 2

### PR DESCRIPTION
**Pull request by le1nux**
# What does this PR do?
Previously, the `block_size` in the dataset would be set to a power of two, resulting in the sequence length being `block_size -1`, which is not best practice and can impact the model training e.g., throughput-wise. 

As a fix, we now specify the `sequence_length` in the config instead of the `block_size`. During Dataset instantiation we chose the `block_size` to be `sequence_length+1`.

Previously, we would also chunk the dataset into `block_size` long chunks. Each chunk would then be used for training individually. As a result, the last token of a block would be only used as a target but never as an input. We changed this, such that we reuse the last token of a batch as the first one of the subsequent batch. 

## General changes
* nothing apart from points mentioned above

## Breaking Changes
* replaced `block_size` in `Dataset`, `Model` and `NumberConversion` with `sequence_length`

## Checklist before submitting final PR
- [x] My PR is minimal and addresses one issue / enhancement in isolation
- [ ] I have merged main into this feature branch
- [x] I have reviewed my own code w.r.t. correct implementation, missing type hints, proper documentation, etc.
- [x] I have run a sample config for model training
- [x] I have fixed all failing tests (`python tests/tests.py`)

### CODE REVIEW

File: src/modalities/models/gpt2/gpt2_model.py [hunk 1]
```diff
@@ -519,7 +518,9 @@ def forward_impl(self, inputs: Dict[str, torch.Tensor]) -> Dict[str, torch.Tenso
         input_ids = inputs[self.sample_key]
         device = input_ids.device
         b, t = input_ids.size()  # batch size, sequence length
-        assert t <= self.block_size, f"Cannot forward sequence of length {t}, block size is only {self.block_size}"
+        assert (
+            t <= self.sequence_length
+        ), f"Cannot forward sequence of length {t}, block size is only {self.sequence_length}"
```
Comment (flxst) [inline src/modalities/models/gpt2/gpt2_model.py:523]:
```suggestion
        ), f"Cannot forward sequence of length {t}, the model's maximum input sequence length is only {self.sequence_length}"
```

File: src/modalities/models/gpt2/gpt2_model.py [hunk 2]
```diff
@@ -519,7 +518,9 @@ def forward_impl(self, inputs: Dict[str, torch.Tensor]) -> Dict[str, torch.Tenso
         input_ids = inputs[self.sample_key]
         device = input_ids.device
         b, t = input_ids.size()  # batch size, sequence length
```
Comment (flxst) [inline src/modalities/models/gpt2/gpt2_model.py:520]:
```suggestion
        _, t = input_ids.size()  # batch size, sequence length
```
batch size `b` is not used

File: tests/checkpointing/test_fsdp_to_disc_checkpointing.py [hunk 3]
```diff
@@ -111,9 +111,9 @@ def _clone_parameters(fsdp_wrapped_model):
     def _generate_batch(gpt2_model_config: Dict):
         # prepare input and targets
         data = torch.randint(
-            0,
-            gpt2_model_config["model"]["config"]["vocab_size"],
-            (8, gpt2_model_config["model"]["config"]["block_size"] + 1),
+            0,  # lowest token_id
+            gpt2_model_config["model"]["config"]["vocab_size"],  # highest token_id, i.e, vocab_size
```
Comment (flxst) [inline tests/checkpointing/test_fsdp_to_disc_checkpointing.py:115]:
```suggestion
            gpt2_model_config["model"]["config"]["vocab_size"],  # highest token_id + 1, i.e. vocab_size
```

**Review by flxst — APPROVED** [review]
Great work! Added some minor comments.

**Review by mali-git — APPROVED** [review]

**Pull request merged**